### PR TITLE
Workaround for correctly exporting IPA when generic archive Xcodebuild bug occurs

### PR DESCRIPTION
**Pull request by jcampbell05**
Workaround for https://github.com/CocoaPods/CocoaPods/issues/4178

In certain project configurations Xcodebuild creates a generic archive with the iOS App with Apple Watch App and the Apple Watch app in the products folder.

This is a small tweak to remove any Watch IPAs in the root of the archive (i.e not inside of a iOS App) since in pretty much all cases a developer wouldn't want / need this.

Hopefully this will increase the chances that `gym` will pick the correct IPA file when this bug occurs.

### CODE REVIEW

**Comment by samrobbins** [comment]
@jcampbell05 what is your sense of how wide spread of a problem this is? Are others seeing the issue too and just not reporting it? Should we put an opt-in mechanism for this? I just don't want to break a workflow for someone that needs it to work this way currently.

File: gym/lib/gym/generators/package_command_generator_xcode7.rb [hunk 1]
```diff
@@ -58,6 +60,7 @@ def ipa_path
             ErrorHandler.handle_empty_archive unless path
           end
         end
+
```
Comment (samrobbins) [inline gym/lib/gym/generators/package_command_generator_xcode7.rb]: Remove?

File: gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      #
+      # This is a workaround for this bug
+      # https://github.com/CocoaPods/CocoaPods/issues/4178
+      def generic_archive_fix
+        UI.verbose "Removing Orphaned WatchKit2 Applications"
+
+        Dir.glob("#{BuildCommandGenerator.archive_path}/Products/Applications/*.app").each do |app_path|
+          if is_watchkit_ipa?("#{app_path}/info.plist")
+            FileUtils.rm_rf(app_path)
+            UI.verbose "Removing #{app_path}"
```
Comment (mfurtak) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: Nitpick: change to "Removed..." or reorder the statements :+1:

File: gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      # Determine IPAs for the Watch App which aren't inside of a containing
+      # iOS App and removes them.
+      #
+      # In the future it may be nice to modify the plist file for the archive
+      # itself so that it points to the correct IPA as well.
+      #
+      # This is a workaround for this bug
+      # https://github.com/CocoaPods/CocoaPods/issues/4178
+      def generic_archive_fix
+        UI.verbose "Removing Orphaned WatchKit2 Applications"
```
Comment (mfurtak) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: Nitpick: change to sentence case - "Removing orphaned WatchKit2 applications"

Comment (samrobbins) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: Perhaps change this to "Looking Orphaned WatchKit2 Applications" and then inside the if block below you can log "Removing Orphaned WatchKit2 Application #{app_path}".

File: gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+          if is_watchkit_ipa?("#{app_path}/info.plist")
+            FileUtils.rm_rf(app_path)
+            UI.verbose "Removing #{app_path}"
+          end
+        end
+      end
+
+      # Does this application have a WatchKit target
+      def is_watchkit_ipa?(plist_path)
+        `/usr/libexec/PlistBuddy -c 'Print DTSDKName' '#{plist_path}' 2>&1`.match(/^\s*watchos2\.\d+\s*$/) != nil
```
Comment (mfurtak) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: For getting the path into the command, do `#{plist_path.shellescape}` without the quotes :+1:

File: gym/spec/xcodebuild_fixes/generic_archive_fix_spec.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    let (:watch_ipa) { 'spec/fixtures/xcodebuild_fixes/watch_app.app/info.plist' }
+    let (:ios_ipa) { 'spec/fixtures/xcodebuild_fixes/ios_app.app/info.plist' }
+
+    it "can detect watch application" do
+      expect(Gym::XcodebuildFixes.is_watchkit_ipa?(watch_ipa)).to eq(true)
+    end
+
+    it "doesn't detect iOS application" do
+      expect(Gym::XcodebuildFixes.is_watchkit_ipa?(ios_ipa)).to eq(false)
+    end
```
Comment (mfurtak) [inline gym/spec/xcodebuild_fixes/generic_archive_fix_spec.rb]: :heart:

**Comment by jcampbell05** [comment]
@samrobbins  This is a pretty wide-spread problem based on the issue filed with CocoaPods https://github.com/CocoaPods/CocoaPods/issues/4178. 

They may not have reported it since potentially it could be something wrong with their project but currently there isn't a straight forward reason to why this bug occurs (But I am working with CocoaPods to identify this).

That said, I am not again putting some sort of flag for now to switch this fix on or off depending on your needs if people need to archive the watch app itself.

The only reason we may not need this flag, is the only useful way to use a watch app archive is inside of a iOS App so potentially this use case is small and most developers would just archive the iOS App.

File: gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb (hunk 4 above)
Comment (jcampbell05) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: @mfurtak  I'll update this but we should think about having utility commands for these things in another PR, this is used elsewhere without `shellescape` unfortunately :(

**Comment by samrobbins** [comment]
@jcampbell05 I have reservations about the assumptions we are making with regard to identifying a project as iOS, tvOS etc. But after I asked the question I thought about it some more and perhaps having things work a certain way vs needing to know extra flags is better. @mfurtak thoughts on this last statement?

**Comment by mfurtak** [comment]
`shellescape` can be brought in with a `require 'shellwords'` :+1:

File: gym/lib/gym/runner.rb [hunk 6]
```diff
@@ -72,6 +73,11 @@ def clear_old_files
       end
     end
 
+    def fix_archive
+      return unless Gym.config[:use_legacy_build_api]
```
Comment (mfurtak) [inline gym/lib/gym/runner.rb]: Let's do as @samrobbins suggests and put this behavior behind a feature switch for now.

Something like:
`return unless Gym.config[:use_legacy_build_api] && !ENV['FASTLANE_GYM_USE_GENERIC_ARCHIVE_FIX'].nil?`

Comment (KrauseFx) [inline gym/lib/gym/runner.rb]: This is confusing, with a lot of `!`, `unless` and `nil?`. Could we split it up to multiple lines?

File: gym/lib/gym/runner.rb [hunk 7]
```diff
@@ -72,6 +73,11 @@ def clear_old_files
...
+      Gym::XcodebuildFixes.generic_archive_fix
```
Comment (KrauseFx) [inline gym/lib/gym/runner.rb]: I'd also really like this to be a feature flag that the user has to enable. Or is there a way to detect if we actually need to apply the fix?

File: gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      # In the future it may be nice to modify the plist file for the archive
+      # itself so that it points to the correct IPA as well.
+      #
+      # This is a workaround for this bug
+      # https://github.com/CocoaPods/CocoaPods/issues/4178
+      def generic_archive_fix
+        UI.verbose "Looking For Orphaned WatchKit2 Applications"
+
+        Dir.glob("#{BuildCommandGenerator.archive_path}/Products/Applications/*.app").each do |app_path|
+          if is_watchkit_ipa?("#{app_path}/info.plist")
```
Comment (KrauseFx) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: Why is this lower case?

Comment (jcampbell05) [inline gym/lib/gym/xcodebuild_fixes/generic_archive_fix.rb]: No reason, I just thought info.plist was lowercase.

File: gym/lib/gym/runner.rb (hunk 7 above)
Comment (jcampbell05) [inline gym/lib/gym/runner.rb]: Not yet until we figure out what causes it. I'll put it behind a flag and in the future if we figure out what causes it I can remove the flag and detect it.

File: gym/lib/gym/runner.rb (hunk 6 above)
Comment (jcampbell05) [inline gym/lib/gym/runner.rb]: Sure :)

**Comment by jcampbell05** [comment]
Updated and all the raised issues resolved.

File: gym/lib/gym/options.rb [hunk 9]
```diff
@@ -103,6 +103,16 @@ def self.plain_options
                                          UI.important "Using legacy build system - waiting for radar to be fixed: https://openradar.appspot.com/radar?id=4952000420642816"
                                        end
                                      end),
+       FastlaneCore::ConfigItem.new(key: :use_generic_archive_fix,
+                                    env_name: "GYM_USE_GENERIC_ARCHIVE_FIX",
+                                    description: "Use fix for generic arhives https://github.com/CocoaPods/CocoaPods/issues/4178",
```
Comment (KrauseFx) [inline gym/lib/gym/options.rb]: There is a spelling mistake `arhives`, could you submit a PR fixing it, or was it already fixed?

**Pull request closed without merging**